### ISSUE TITLE
Broken SSR: You should not use <Link> outside a <Router>

### ISSUE THREAD

**Issue by prontiol** [ReactTraining/react-router#6704]
We're developing a SPA application with SSR and our SSR broke after update to v5.0.0.
Everything worked fine in v4.3.1 and broke because of switching to the new React Context API.
We have no problems on client side with `react-router` v5.0.0 as all our dynamically loading chunks are sharing `react-router` and `react-router-dom` packages (via [webpack dll](https://webpack.js.org/plugins/dll-plugin/)), but for SSR we render those chunks separately and they do not share any packages, they are fully standalone chunks which are able to render themselves.
So here comes the problem: there is an `invariant` check inside the router and it fails because chunks expect context from their own imported `react-router` which is obviously not there - only our main chunk renders the `<Router>` and children chunks can't import `react-router` from its bundle.
This invariant is *not* needed to make `react-router` work, this is just a utility to help people identify router problems, but it totally breaks our SSR.
Is there a way we can disable this invariant without forking a whole package? Maybe we can remove it or make it less destructive (e.g. output a warning instead)?
Thanks.

**Comment by StringEpsilon** [comment]
> This invariant is not needed to make react-router work

For some components (mainly Link and NavLink) you could implement an alternative render path in case the context is missing. But for Route, Switch, Redirect and Prompt, there is nothing meaningful you can do in case the context is empty and or missing.

I suppose you could log a warning and return null. 

BTW: You might want to reformat your issue to conform to the issue template. A small reproduction repository / sandbox would also be greatly appreciated. 

On a last note: Is there a way to detect SSR? I fear that some changes required for your scenario would be considered breaking and require a major version bump. Mainly because `withRouter()` would no longer be guaranteed to supply `match`, `history` and `location`. Having that new behavior only in SSR would lessen the impact of the change.

**Comment by prontiol** [comment]
> For some components (mainly Link and NavLink) you could implement an alternative render path in case the context is missing.

I don't really want to introduce workarounds to avoid diagnostic code (which invariant is) in `react-router`. This seems to be fragile as `react-router` can add more invariants later on, forcing me to add more and more workarounds. This will also bloat the code.
I would suggest putting those invariants into `componentDidMount` instead of `render` so they will not be called during the SSR.

> Edit: Also, do I understand you correctly that you're importing from both react-router and react-router-dom? I'd advice against that. Some people here and on reactiflux solved their invariant exceptions by changing all imports to react-router-dom.

This won't help us, as we have separate bundles and `react-router-dom` is not shared as well as `react-router` itself.

Basically, react-router broke some use cases just to output more human-readable breaking error (using invariant) instead of javascript generic error. I don't think it worth it.

**Comment by StringEpsilon** [comment]
The breakage comes from using the new context, not from the invariant. RR 4.3 also warns you if components are used outside a <Router/>, but it used the old context system.

> I don't really want to introduce workarounds to avoid diagnostic code (which invariant is) in react-router. 
> I would suggest putting those invariants into componentDidMount instead of render so they will not be called during the SSR.

I tried to tell you that the invariant is not diagnostic in nature. It's an assert on a criticical precondition. Without a context, Link and NavLink as they are currently implemented won't work. If we could move the invariant to `componentWillMount` (which would require mandating react 16.4 -> breaking change), NavLink and Link would simply throw null reference exceptions or other errors. We'd have to bail out of rendering by returning null or having an alternate path that doesn't rely on context but still makes the components output useful.

**Comment by DenysMikhalenkoWalmart** [comment]
I still believe we can render `<Link>`s properly without context as it will be needed only in `onClick` handler which will never be called in SSR.

**Comment by DenysMikhalenkoWalmart** [comment]
Okay, I assume this is more React Context API problem than RR's one.

**Comment by prontiol** [comment]
Ok, I assume this is more React Context API problem than RR's one.
I am closing this issue.
